processing/src/test/java/org/apache/druid/segment/data/VSizeLongSerdeTest.java

```diff
@@ -20,132 +20,352 @@
 package org.apache.druid.segment.data;
 
 
+import com.google.common.primitives.Ints;
+import org.apache.druid.java.util.common.StringUtils;
 import org.junit.Assert;
-import org.junit.Before;
 import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.stream.Collectors;
 
+@RunWith(Enclosed.class)
 public class VSizeLongSerdeTest
```

Comment:
The change in Mult4Ser suggests that we care about handling negative numbers, but this test class doesn't exercise negative numbers very much. (I think it only tests Long.MIN_VALUE, in testEveryPowerOfTwo.)

If negative numbers matter, we should extend the test cases in this file to cover them better. I'd suggest adding tests to EveryLittleBitTest that are similar to testEveryPowerOfTwo and testEveryPowerOfTwoMinusOne, but have the sign bit set (i.e. bitwise or with `Long.MIN_VALUE`).

If negative numbers aren't important, I'd suggest blocking them on the write side, i.e. have all the LongSerializers throw errors if they are fed negative numbers.

Reply:
I don't think its necessary since that change was accidental, and I think negative numbers shouldn't make it here because both delta and table encoding appear to make all numbers positive.

Reply:
In that case, IMO it'd be good to add checks to the serializers to make sure negative numbers aren't provided.

Reply:
Looking closer, I don't think it is necessary to add the checks based on how `IntermediateColumnarLongsSerializer` works. 

For delta encoding, the delta is computed with `LongMath.checkedSubtract` which will ensure delta is a positive number and the checks ensure it won't be used if delta is `Long.MAX_VALUE`, or it overflowed because the range between values was too big, setting delta to -1. These checks make it so the value in the constructor of the writer fed to `VSizeLongSerde.getBitsForMax` will be appropriate. The `minVal` is used as the base, and the writer subtracts the base from every value while serializing so every number given to the serializer will be between 0 and delta.

For table encoding, since the values are replaced with their corresponding index into the table array, they should also always be 0 or positive.

Is it worth the overhead to make `LongSerializer.write` implementations check this for every value?

Reply:
It doesn't really matter much if the methods are being called correctly today. The purpose of the checks is in case someone starts calling these methods incorrectly in the future.

I would imagine the overhead is minimal given that LongSerializers are doing a single write per method call. (A precondition check shouldn't add much on top of the method call overhead that's already there, especially since the branch would be 100% predictable.)

If it turns out that the overhead is worth worrying about, then I'd add an assert instead of a regular precondition check. The assert would be skipped in production but will run during tests.

Reply:
It's not too bad, low end doesn't seem very different at all

Before adding `Preconditions.checkArgument(value >= 0);` to every write:
```
Benchmark                                                                   (distribution)  (encoding)   (rows)  (zeroProbability)  Mode  Cnt         Score   Error  Units
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-1    lz4-auto  5000000                0.0  avgt    2        97.509          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-2    lz4-auto  5000000                0.0  avgt    2       148.329          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-4    lz4-auto  5000000                0.0  avgt    2       171.316          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-8    lz4-auto  5000000                0.0  avgt    2       242.792          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-12    lz4-auto  5000000                0.0  avgt    2       249.412          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-16    lz4-auto  5000000                0.0  avgt    2       254.304          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-20    lz4-auto  5000000                0.0  avgt    2       309.736          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-24    lz4-auto  5000000                0.0  avgt    2       359.252          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uinform-32    lz4-auto  5000000                0.0  avgt    2       416.235          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-40    lz4-auto  5000000                0.0  avgt    2       492.306          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-48    lz4-auto  5000000                0.0  avgt    2       611.416          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-56    lz4-auto  5000000                0.0  avgt    2       734.503          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-64    lz4-auto  5000000                0.0  avgt    2       669.983          ms/op
```

after:
```
Benchmark                                                                   (distribution)  (encoding)   (rows)  (zeroProbability)  Mode  Cnt         Score   Error  Units
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-1    lz4-auto  5000000                0.0  avgt    2       100.281          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-2    lz4-auto  5000000                0.0  avgt    2       148.947          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-4    lz4-auto  5000000                0.0  avgt    2       178.900          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                       uniform-8    lz4-auto  5000000                0.0  avgt    2       264.585          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-12    lz4-auto  5000000                0.0  avgt    2       262.852          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-16    lz4-auto  5000000                0.0  avgt    2       265.252          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-20    lz4-auto  5000000                0.0  avgt    2       341.000          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-24    lz4-auto  5000000                0.0  avgt    2       368.859          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uinform-32    lz4-auto  5000000                0.0  avgt    2       456.571          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-40    lz4-auto  5000000                0.0  avgt    2       635.758          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-48    lz4-auto  5000000                0.0  avgt    2       831.513          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-56    lz4-auto  5000000                0.0  avgt    2       975.933          ms/op
ColumnarLongsEncodeDataFromGeneratorBenchmark.encodeColumn                      uniform-64    lz4-auto  5000000                0.0  avgt    2       727.638          ms/op
```

Reply:
oops, preconditions broke test for 64 bit `testEveryPowerOfTwo`, which overflows, though thinking about it, I guess it wouldn't matter if 64 is negative because it has all the bits. I've made an exception for this case to make this test pass again.